### PR TITLE
fix: user deletion module update

### PR DESCRIPTION
**Pull request by anantjain45823**
## Description of the change

> Fixing the user deletion module with mocking and refactoring where needed.
Changed the API for Intercom as per their documentation.
Supported Batching in Braze and Amplitude
Changed the Batching Method for Sendgrid using the URL Limit from Max ids allowed per batch since it's not given.

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [x] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [x] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

**Review by saikumarrs — CHANGES_REQUESTED** [review]
Partial review 1. Probably due to cherry-picking commits, some recent changes were overridden. Please take care of that.

File: src/v0/destinations/am/deleteUsers.js [hunk 1]
```diff
@@ -1,17 +1,19 @@
...
 const { isHttpStatusSuccess } = require("../../util");
 const {
+  ConfigurationError,
   NetworkError,
-  InstrumentationError,
-  ConfigurationError
+  InstrumentationError
 } = require("../../util/errorTypes");
 const { executeCommonValidations } = require("../../util/regulation-api");
-const tags = require("../../util/tags");
```
Comment (saikumarrs) [inline src/v0/destinations/am/deleteUsers.js:14]: This import is not supposed to be deleted.

**Review by saikumarrs — CHANGES_REQUESTED** [review]

File: src/v0/destinations/sendgrid/deleteUsers.js [hunk 2]
```diff
@@ -13,9 +12,38 @@ const {
...
+ * @returns list of Strings
+ */
+const chunksFromUrlLength = (userAttributes, maxSize) => {
+  const identity = [];
+  let left = maxSize;
+  let temp = "";
+  userAttributes.forEach(ua => {
+    // Dropping the user if userId is not present
+    if (ua.userId) {
+      left -= String(ua.userId).length;
```
Comment (saikumarrs) [inline src/v0/destinations/sendgrid/deleteUsers.js:29]: Shouldn't we include the comma `,` also in the length computation?

File: src/v0/destinations/sendgrid/deleteUsers.js [hunk 3]
```diff
@@ -13,9 +12,38 @@ const {
...
 const tags = require("../../util/tags");
 
+/**
+ * This drops the user if userId is not available and converts the ids's into list of strings
+ * where each string is a combination of comma separated userIds and length of each string is not more than maxSize
+ * @param {*} userAttributes array of userIds
+ * @param {*} maxSize maxSize of url
+ * @returns list of Strings
+ */
+const chunksFromUrlLength = (userAttributes, maxSize) => {
```
Comment (saikumarrs) [inline src/v0/destinations/sendgrid/deleteUsers.js:22]:
```suggestion
const getUserIdChunks = (userAttributes, maxUrlLength) => {
```

File: src/v0/destinations/sendgrid/config.js [hunk 4]
```diff
@@ -38,8 +38,9 @@ const TRACK_EXCLUSION_FIELDS = [
   "trackingSettings"
 ];
 
-const DELETE_CONTACTS_ENDPOINT = `${BASE_ENDPOINT}/marketing/contacts`;
+const DELETE_CONTACTS_ENDPOINT = `${BASE_ENDPOINT}/marketing/contacts?ids=[IDS]`;
 const MAX_BATCH_SIZE = 30000;
+const urlLimit = 8150;
```
Comment (saikumarrs) [inline src/v0/destinations/sendgrid/config.js:43]: Is this the limit for the entire URL or just the `ids` query parameter?

File: src/v0/destinations/sendgrid/config.js [hunk 5]
```diff
@@ -38,8 +38,9 @@ const TRACK_EXCLUSION_FIELDS = [
   "trackingSettings"
 ];
 
-const DELETE_CONTACTS_ENDPOINT = `${BASE_ENDPOINT}/marketing/contacts`;
+const DELETE_CONTACTS_ENDPOINT = `${BASE_ENDPOINT}/marketing/contacts?ids=[IDS]`;
 const MAX_BATCH_SIZE = 30000;
```
Comment (saikumarrs) [inline src/v0/destinations/sendgrid/config.js:42]: This is not required, right?

File: src/v0/destinations/sendgrid/config.js (hunk 4 above)
Comment (saikumarrs) [inline src/v0/destinations/sendgrid/config.js:43]: Not that there is a limit of max URL length in HTTP requests in general.

File: src/v0/destinations/mp/deleteUsers.js [hunk 6]
```diff
@@ -41,13 +45,16 @@ const userDeletionHandler = async (userAttributes, config) => {
       });
     }
   });
+  if (data.length === 0) {
+    throw new InstrumentationError(`No User id for deletion is present`);
+  }
```
Comment (saikumarrs) [inline src/v0/destinations/mp/deleteUsers.js:50]: Remove this.

File: src/v0/destinations/mp/deleteUsers.js [hunk 7]
```diff
@@ -6,13 +6,17 @@ const {
...
+  NetworkError,
+  InstrumentationError
+} = require("../../util/errorTypes");
 const { getDynamicErrorType } = require("../../../adapters/utils/networkUtils");
 const tags = require("../../util/tags");
 
 /**
  * This function will help to delete the users one by one from the userAttributes array.
- * @param {*} userAttributes Array of objects with userId, email and phone
+ * @param {*} userAttributes Array of objects with userId, emaail and phone
```
Comment (saikumarrs) [inline src/v0/destinations/mp/deleteUsers.js:19]:
```suggestion
 * @param {*} userAttributes Array of objects with userId, email and phone
```

**Review by saikumarrs — CHANGES_REQUESTED** [review]

File: src/v0/destinations/sendgrid/deleteUsers.js (hunk 2 above)
Comment (anantjain45823) [reply src/v0/destinations/sendgrid/deleteUsers.js:29]: Yes, We should. incorporating it.

File: src/v0/destinations/sendgrid/config.js (hunk 4 above)
Comment (anantjain45823) [reply src/v0/destinations/sendgrid/config.js:43]: Only the ids.

File: src/v0/destinations/sendgrid/config.js (hunk 5 above)
Comment (anantjain45823) [reply src/v0/destinations/sendgrid/config.js:42]: `MAX_BATCH_SIZE` is required in tranform.js of sendgrid

File: src/v0/destinations/sendgrid/config.js (hunk 4 above)
Comment (anantjain45823) [reply src/v0/destinations/sendgrid/config.js:43]: Yeah 2048. But through the trial and error method I found it to be this.

Comment (saikumarrs) [reply src/v0/destinations/sendgrid/config.js:43]: Then we should clarify that in the name. Currently, it is ambiguous.

File: src/v0/destinations/intercom/deleteUsers.js [hunk 8]
```diff
@@ -24,32 +21,34 @@ const userDeletionHandler = async (userAttributes, config) => {
   await Promise.all(
     userAttributes.map(async ua => {
```
Comment (saikumarrs) [inline src/v0/destinations/intercom/deleteUsers.js:22]: Filter the valid user IDs first here.

File: src/v0/destinations/am/deleteUsers.js [hunk 9]
```diff
@@ -21,35 +23,47 @@ const userDeletionHandler = async (userAttributes, config) => {
...
+          )}`,
+          handledResponse.status,
           {
             [tags.TAG_NAMES.ERROR_TYPE]: getDynamicErrorType(
-              handledDelResponse.status
+              handledResponse.status
             )
           },
-          handledDelResponse
+          handledResponse
```
[the remark this answers is not in the document]
Comment (anantjain45823) [reply src/v0/destinations/am/deleteUsers.js]: Done

File: src/v0/destinations/am/deleteUsers.js [hunk 10]
```diff
@@ -21,35 +23,47 @@ const userDeletionHandler = async (userAttributes, config) => {
...
-        throw new InstrumentationError("User id for deletion not present");
-      }
-      const data = { user_ids: [uId], requester: "RudderStack" };
+    batchEvents.map(async batch => {
+      const data = {
+        user_ids: batch,
+        requester: "RudderStack",
+        ignore_invalid_id: "true"
+      };
+      const url = "https://amplitude.com/api/2/deletions/users";
```
[the remark this answers is not in the document]
Comment (anantjain45823) [reply src/v0/destinations/am/deleteUsers.js]: Done

File: src/v0/destinations/am/config.js [hunk 11]
```diff
@@ -118,6 +118,7 @@ const events = Object.keys(Event);
 events.forEach(event => {
   nameToEventMap[Event[event].name] = Event[event];
 });
+const DELETE_MAX_BATCH_SIZE = 100;
```
[the remark this answers is not in the document]
Comment (anantjain45823) [reply src/v0/destinations/am/config.js:122]: Done

File: src/v0/destinations/braze/config.js [hunk 12]
```diff
@@ -29,7 +29,7 @@ const BRAZE_PARTNER_NAME = "RudderStack";
 // Ref: https://www.braze.com/docs/api/endpoints/user_data/post_user_track/
 const TRACK_BRAZE_MAX_REQ_COUNT = 75;
 const IDENTIFY_BRAZE_MAX_REQ_COUNT = 50;
-
+const MAX_BATCH_SIZE = 50;
```
[the remark this answers is not in the document]
Comment (anantjain45823) [reply src/v0/destinations/braze/config.js]: Done

File: src/v0/destinations/clevertap/deleteUsers.js [hunk 13]
```diff
@@ -38,7 +42,9 @@ const userDeletionHandler = async (userAttributes, config) => {
       identity.push(userAttribute.userId);
     }
   });
-
+  if (identity.length === 0) {
+    throw new InstrumentationError(`No User id for deletion is present`);
+  }
```
[the remark this answers is not in the document]
Comment (anantjain45823) [reply src/v0/destinations/clevertap/deleteUsers.js]: Done

File: src/v0/destinations/braze/deleteUsers.js [hunk 14]
```diff
@@ -22,50 +24,55 @@ const userDeletionHandler = async (userAttributes, config) => {
...
+    if (userAttribute.userId) {
+      identity.push(userAttribute.userId);
+    }
+  });
+  if (identity.length === 0) {
+    throw new InstrumentationError(`No User id for deletion is present`);
+  }
+  // batchEvents = [[e1,e2,e3,..batchSize],[e1,e2,e3,..batchSize]..]
+  // https://www.braze.com/docs/api/endpoints/user_data/post_user_delete#request-body
+  const batchEvents = _.chunk(identity, MAX_BATCH_SIZE);
```
[the remark this answers is not in the document]
Comment (anantjain45823) [reply src/v0/destinations/braze/deleteUsers.js]: Done

File: src/v0/destinations/braze/deleteUsers.js [hunk 15]
```diff
@@ -22,50 +24,55 @@ const userDeletionHandler = async (userAttributes, config) => {
...
       if (
-        !isHttpStatusSuccess(handledDelResponse.status) &&
-        handledDelResponse.status !== 404
+        !isHttpStatusSuccess(handledResponse.status) &&
+        handledResponse.status !== 404
       ) {
         throw new NetworkError(
-          "User deletion request failed",
-          handledDelResponse.status,
+          `user deletion request failed - error: ${JSON.stringify(
```
[the remark this answers is not in the document]
Comment (anantjain45823) [reply src/v0/destinations/braze/deleteUsers.js]: Done

File: src/v0/destinations/engage/deleteUsers.js [hunk 16]
```diff
@@ -1,20 +1,20 @@
...
+  InstrumentationError,
+  ConfigurationError,
+  NetworkError
+} = require("../../util/errorTypes");
 const { executeCommonValidations } = require("../../util/regulation-api");
 
 /**
  * This function will help to delete the users one by one from the userAttributes array.
- * @param {*} userAttributes Array of objects with userId, email and phone
+ * @param {*} userAttributes Array of objects with userId, emaail and phone
```
[the remark this answers is not in the document]
Comment (anantjain45823) [reply src/v0/destinations/engage/deleteUsers.js]: Done

File: src/v0/destinations/intercom/deleteUsers.js [hunk 17]
```diff
@@ -25,30 +25,35 @@ const userDeletionHandler = async (userAttributes, config) => {
...
-        url: `https://api.intercom.io/contacts/${uId}`,
         headers: {
-          Authorization: `Bearer ${apiKey}`
+          Authorization: `Bearer ${apiKey}`,
+          Accept: "application/json"
         }
       };
-      const resp = await httpSend(requestOptions);
-      const handledDelResponse = processAxiosResponse(resp);
+      const resp = await httpPOST(url, data, requestOptions);
```
[the remark this answers is not in the document]
Comment (anantjain45823) [reply src/v0/destinations/intercom/deleteUsers.js:39]: Yeah it is according to their latest Doc.Ref https://developers.intercom.com/intercom-api-reference/v1.3/reference/permanently-delete-a-user

File: src/v0/destinations/am/deleteUsers.js (hunk 1 above)
Comment (anantjain45823) [reply src/v0/destinations/am/deleteUsers.js:14]: Removed

File: src/v0/util/deleteUserUtils.js [hunk 18]
```diff
@@ -0,0 +1,21 @@
+const _ = require("lodash");
+/**
+ * Takes the userAttributes array fetches userIds at root level of each element
+ * and make batches of Ids accoring to MAX_BATCH_SIZE
+ * @param {*} userAttributes array
+ * @param {*} MAX_BATCH_SIZE integer
+ * @returns [e1,e2,e3,..batchSize],[e1,e2,e3,..batchSize]..]
+ */
+const getBatchedIds = (userAttributes, MAX_BATCH_SIZE) => {
```
[the remark this answers is not in the document]
Comment (anantjain45823) [reply src/v0/util/deleteUserUtils.js]: Done

File: src/v0/util/deleteUserUtils.js [hunk 19]
```diff
@@ -0,0 +1,21 @@
...
+  const identity = [];
```
[the remark this answers is not in the document]
Comment (anantjain45823) [reply src/v0/util/deleteUserUtils.js]: Done

File: src/v0/util/deleteUserUtils.js [hunk 20]
```diff
@@ -0,0 +1,21 @@
...
+  userAttributes.forEach(userAttribute => {
+    // Dropping the user if userId is not present
+    if (userAttribute.userId) {
+      identity.push(userAttribute.userId);
+    }
+  });
+  const batchEvents = _.chunk(identity, MAX_BATCH_SIZE);
+  return batchEvents;
```
[the remark this answers is not in the document]
Comment (anantjain45823) [reply src/v0/util/deleteUserUtils.js]: Done

File: src/v0/destinations/sendgrid/deleteUsers.js (hunk 2 above)
Comment (anantjain45823) [reply src/v0/destinations/sendgrid/deleteUsers.js]: Changed the flow. Please check

File: src/v0/destinations/sendgrid/deleteUsers.js (hunk 3 above)
Comment (anantjain45823) [reply src/v0/destinations/sendgrid/deleteUsers.js]: Done

File: src/v0/destinations/mp/deleteUsers.js (hunk 6 above)
Comment (anantjain45823) [reply src/v0/destinations/mp/deleteUsers.js]: Done

File: src/v0/destinations/mp/deleteUsers.js (hunk 7 above)
Comment (anantjain45823) [reply src/v0/destinations/mp/deleteUsers.js]: Done

File: src/v0/destinations/intercom/deleteUsers.js (hunk 8 above)
Comment (anantjain45823) [reply src/v0/destinations/intercom/deleteUsers.js]: Done

File: src/v0/destinations/sendgrid/config.js (hunk 4 above)
Comment (anantjain45823) [reply src/v0/destinations/sendgrid/config.js]: Done

File: src/v0/destinations/sendgrid/deleteUsers.js [hunk 21]
```diff
@@ -13,6 +12,40 @@ const {
...
+  userAttributes.forEach(ua => {
+    // Dropping the user if userId is not present
+    if (ua.userId) {
+      idBatch.push(ua.userId);
+      if (idBatch.toString().length < maxUrlLength) {
+        idBatchString = idBatch.toString();
+      } else {
+        userIdBatchEndpoints.push(
+          `${DELETE_CONTACTS_ENDPOINT.replace("IDS", idBatchString)}`
+        );
```
[the remark this answers is not in the document]
Comment (anantjain45823) [reply src/v0/destinations/sendgrid/deleteUsers.js]: Done

**Review by saikumarrs — APPROVED** [review]

**Pull request merged**